### PR TITLE
Refactors the scanPokemon method, build on top of blaisorblades crashfix

### PR DESCRIPTION
**Pull request by nahojjjen**
@Blaisorblade made a fix for the crash on scan failure, I ontop of that refactored the methods, this pull request includes both his and my commits.

### CODE REVIEW

File: app/src/main/java/com/kamron/pogoiv/MainActivity.java [hunk 1]
```diff
@@ -587,12 +586,14 @@ private void takeScreenshot() {
...
 
+
     /**
      * scans the arc and tries to determine the pokemon level, returns 1 if nothing found
      * @param pokemonImage The image of the entire screen
      * @return the estimated pokemon level, or 1 if nothing found
      */
-    private double getPokemonLevel(Bitmap pokemonImage){
+    private double getPokemonLevelFromImg(Bitmap pokemonImage, double trainerLevel){
+        double estimatedPokemonLevel = trainerLevel + 1.5;
```
Comment (Blaisorblade) [inline app/src/main/java/com/kamron/pogoiv/MainActivity.java]: 👍 , this looks good :-).

File: app/src/main/java/com/kamron/pogoiv/MainActivity.java [hunk 2]
```diff
@@ -717,18 +718,21 @@ private void scanPokemon(Bitmap pokemonImage, String filePath) {
...
-        int pokemonHP = getPokemonHPFromImg(pokemonImage);
-        int pokemonCP = getPokemonCPFromImg(pokemonImage);
+        try {
+            double estimatedPokemonLevel = getPokemonLevelFromImg(pokemonImage, trainerLevel);
+            String pokemonName = getPokemonNameFromImg(pokemonImage);
+            String candyName = getCandyNameFromImg(pokemonImage);
+            int pokemonHP = getPokemonHPFromImg(pokemonImage);
+            int pokemonCP = getPokemonCPFromImg(pokemonImage);
+            if (candyName.equals("") && pokemonHP == 10 && pokemonCP == 10){ //the default values for a failed scan, if all three fail, then probably scrolled down.
+                Toast.makeText(MainActivity.this, "Please scroll up and scan again", Toast.LENGTH_SHORT).show();
```
Comment (Blaisorblade) [inline app/src/main/java/com/kamron/pogoiv/MainActivity.java]: Let's use a resource string so people can translate it. I added scan_pokemon_failed before:
https://github.com/farkam135/GoIV/pull/168/commits/4ee1aa7f031357bbdea8626b6bc8c216419b7be9
but your string might be better.

**Comment by Blaisorblade** [comment]
So this also includes the old version of #152? Hopefully it doesn't matter too much :-)

**Comment by Blaisorblade** [comment]
The base PR is #164, don't need to merge both.

**Comment by jehy** [comment]
Avoid using double when you can use float.

File: app/src/main/java/com/kamron/pogoiv/MainActivity.java [hunk 3]
```diff
@@ -612,96 +616,123 @@ private double getPokemonLevel(Bitmap pokemonImage){
...
+        String candyName ="";
         //System.out.println(tesseract.getUTF8Text());
         //SaveImage(candy, "candy");
         try {
             candyName = tesseract.getUTF8Text().trim().replace("-", " ").split(" ")[candyOrder].replace(" ", "").replace("1", "l").replace("0", "o");
             candyName = new StringBuilder().append(candyName.substring(0, 1)).append(candyName.substring(1).toLowerCase()).toString();
-        }catch(StringIndexOutOfBoundsException e){
-            candyName = pokemonName; //Default for not finding candy name
+        } catch (StringIndexOutOfBoundsException e) {
+            candyName = "";
```
Comment (Blaisorblade) [inline app/src/main/java/com/kamron/pogoiv/MainActivity.java]: This assignment is redundant with the initialization, no?

File: app/src/main/java/com/kamron/pogoiv/MainActivity.java [hunk 4]
```diff
@@ -612,96 +616,123 @@ private double getPokemonLevel(Bitmap pokemonImage){
...
-        //System.out.println(tesseract.getUTF8Text());
         String pokemonHPStr = tesseract.getUTF8Text();
-
-        //Check if valid pokemon TODO find a better method of determining whether or not this is a pokemon image
-        if(pokemonHPStr.contains("/")) {
+        if (pokemonHPStr.contains("/")) {
             try {
                 pokemonHP = Integer.parseInt(pokemonHPStr.split("/")[1].replace("Z", "2").replace("O", "0").replace("l", "1").replaceAll("[^0-9]", ""));
             } catch (java.lang.NumberFormatException e) {
                 pokemonHP = 10;
```
Comment (Blaisorblade) [inline app/src/main/java/com/kamron/pogoiv/MainActivity.java]: This assignment is also redundant with the initialization, and we can't drop the initialization here.

**Comment by Blaisorblade** [comment]
You mean for the level? Speed-wise double and float are apparently equivalent even on ARM, I doubt it's worth changing it everywhere, and I doubt it's worth making this inconsistent.
https://developer.android.com/training/articles/perf-tips.html#AvoidFloat

**Comment by jehy** [comment]
From your link:

> In speed terms, there's no difference between float and double on the more modern hardware. Space-wise, double is 2x larger. As with desktop machines, assuming space isn't an issue, you should prefer double to float.

**Comment by Blaisorblade** [comment]
I've read that. Your point? We're talking about 1 double, not an array.

**Comment by Blaisorblade** [comment]
Meanwhile, I've cherry-picked @nahojjjen's commits on top of #164, to avoid the confusing double commits etc.

**Comment by DJCrapsody** [comment]
From: https://www.reddit.com/r/GoIV/comments/4zarig/bug_if_you_click_the_iv_button_while_the_pokemon/d6urlno

morganga  
The game also crashes if you manually edit the CP and accidentally delete all digits, then press "calculate"

**Comment by nahojjjen** [comment]
Blaisorblade included these changes in #164

**Comment by Blaisorblade** [comment]
@DJCrapsody I filed #171 about that, it's a different bug.

**Comment by DJCrapsody** [comment]
@Blaisorblade have noted. Will be updating the thread on reddit.

**Pull request closed without merging**